src/memory_manager.rs

```diff
@@ -853,4 +1032,104 @@ mod test {
+        assert_eq!(memory_1.grow(BUCKET_SIZE_IN_PAGES), 1);
+        assert_eq!(mem.size(), 1 + BUCKET_SIZE_IN_PAGES * 4);
+        assert_eq!(mem_mgr.get(MemoryId(1)).size(), 1 + BUCKET_SIZE_IN_PAGES);
+
+        // Free Memory ID 1.
+        mem_mgr.free(MemoryId(1));
+        assert_eq!(mem_mgr.get(MemoryId(1)).size(), 0);
+        assert_eq!(mem.size(), 1 + BUCKET_SIZE_IN_PAGES * 4);
+
+        let memory_2 = mem_mgr.get(MemoryId(2));
+        // When growing memory_2, mem.size() should stay the same since
+        // MemoryManager should use the memory that is freed above.
+        assert_eq!(memory_2.grow(1), 0);
+        assert_eq!(mem_mgr.get(MemoryId(2)).size(), 1);
+        assert_eq!(mem.size(), 1 + BUCKET_SIZE_IN_PAGES * 4);
+        assert_eq!(memory_2.grow(BUCKET_SIZE_IN_PAGES), 1);
+        assert_eq!(mem_mgr.get(MemoryId(2)).size(), 1 + BUCKET_SIZE_IN_PAGES);
+        assert_eq!(mem.size(), 1 + BUCKET_SIZE_IN_PAGES * 4);
+
+        // When trying to grow memory_2 again, we need more pages,
+        // because we have already used all that is left from Memory ID 1.
+        assert_ne!(memory_2.grow(BUCKET_SIZE_IN_PAGES), 0);
+        assert_eq!(
+            mem_mgr.get(MemoryId(2)).size(),
+            1 + 2 * BUCKET_SIZE_IN_PAGES
+        );
+        assert_eq!(mem.size(), 1 + BUCKET_SIZE_IN_PAGES * 5);
+    }
+
+    #[test]
+    #[should_panic = "MemoryId(1): read out of bounds"]
+    fn can_be_reinitialized_from_memory_after_freeing() {
+        let mem = make_memory();
+        let mem_mgr = MemoryManager::init(mem.clone());
+        let memory_0 = mem_mgr.get(MemoryId(0));
+        let memory_1 = mem_mgr.get(MemoryId(1));
+        let memory_2 = mem_mgr.get(MemoryId(2));
+
+        assert_eq!(memory_0.grow(1), 0);
+        assert_eq!(memory_1.grow(1), 0);
+        assert_eq!(memory_2.grow(1), 0);
+
+        memory_0.write(0, &[1, 2, 3]);
+        memory_1.write(0, &[4, 5, 6]);
+        memory_2.write(0, &[7, 8, 9]);
+
+        let mut mem_mgr = MemoryManager::init(mem);
+        let memory_0 = mem_mgr.get(MemoryId(0));
+        let memory_1 = mem_mgr.get(MemoryId(1));
+        let memory_2 = mem_mgr.get(MemoryId(2));
+
+        let mut bytes = vec![0; 3];
+        // Check that data is correctly reinitialized.
+        memory_0.read(0, &mut bytes);
+        assert_eq!(bytes, vec![1, 2, 3]);
+
+        memory_1.read(0, &mut bytes);
+        assert_eq!(bytes, vec![4, 5, 6]);
+
+        memory_2.read(0, &mut bytes);
+        assert_eq!(bytes, vec![7, 8, 9]);
+
+        // Free MemoryId 1.
+        mem_mgr.free(MemoryId(1));
+
+        // Check that data of MemoryId 0 and 2 is correctly reinitialized.
+        memory_0.read(0, &mut bytes);
+        assert_eq!(bytes, vec![1, 2, 3]);
+
+        memory_2.read(0, &mut bytes);
+        assert_eq!(bytes, vec![7, 8, 9]);
+
+        // Check that date of MemoryId 1 is freed.
+        assert_eq!(memory_1.size(), 0);
+        memory_1.read(0, &mut bytes);
```

Comment:
Can this test not simply be? It's not clear to me what the additional logic adds to the test.

```
       let mem = make_memory();
       let mem_mgr = MemoryManager::init(mem.clone());
       let memory_0 = mem_mgr.get(MemoryId(0));
       mem_mgr.free(MemoryId(0));
       memory_0.read(0, &mut bytes);
```